[B]Add EntityTeleportEvent cause. Adds BUKKIT-4745

Obliv opened the pull request:
Issue
EntityTeleportEvent does not have any information about teleport cause, while PlayerTeleportEvent does.

Justification
This standardizes the capabilities of teleport events between players and entities. 

PR Breakdown
Added a teleport cause parameter to EntityTeleportEvent. Updated
relevant calls to constructors in child classes.

JIRA Ticket
https://bukkit.atlassian.net/browse/BUKKIT-4745

Obliv committed:
Add EntityTeleportEvent cause. Adds BUKKIT-4745

Added a teleport cause parameter to EntityTeleportEvent. Updated
relevant calls to constructors.

mbax commented on src/main/java/org/bukkit/event/entity/EntityPortalEvent.java:
```diff
@@ -16,8 +17,8 @@
     protected boolean useTravelAgent = true;
     protected TravelAgent travelAgent;
 
-    public EntityPortalEvent(final Entity entity, final Location from, final Location to, final TravelAgent pta) {
```
Must add NEW constructor, deprecating old one.

mbax commented on src/main/java/org/bukkit/event/entity/EntityPortalExitEvent.java:
```diff
@@ -3,21 +3,22 @@
     private Vector before;
     private Vector after;
 
-    public EntityPortalExitEvent(final Entity entity, final Location from, final Location to, final Vector before, final Vector after) {
```
Must add new constructor and deprecate old

mbax commented on src/main/java/org/bukkit/event/entity/EntityTeleportEvent.java:
```diff
@@ -14,12 +15,45 @@
     private Location to;
+    private TeleportCause cause = TeleportCause.UNKNOWN;
 
-    public EntityTeleportEvent(Entity what, Location from, Location to) {
```
Must add new constructor and deprecate old

Obliv committed:
Deprecate constructors

Changed the constructors back, deprecated them, and added the updated
constructors.

Obliv committed:
Fixed whitespace

riking commented on src/main/java/org/bukkit/event/entity/EntityPortalEvent.java:
```diff
@@ -16,11 +17,20 @@
     protected TravelAgent travelAgent;
 
+    /*
+     * @Deprecated Super constructor deprecated: No teleport cause
```
I think we generally don't provide deprecation reasons for event constructors - it's usually pretty dang clear by looking at the file.

Also, it's `@deprecated` not `@Deprecated` in the javadocs.

turt2live replied:
Correct, we do not generally add JavaDocs to deprecated event constructors.

Please remove the JavaDoc portion of this.

Obliv replied:
Alright, I will remove it.

riking commented on src/main/java/org/bukkit/event/entity/EntityTeleportEvent.java:
```diff
@@ -4,6 +4,7 @@
 import org.bukkit.entity.Entity;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
```
Wait, you import a different `TeleportCause` class and define your own? How the heck is this not a compile error?

Obliv replied:
I honestly don't know how this compiled. I didn't notice that Eclipse had added that import. I will remove the duplicate cause and keep this import.

riking commented on src/main/java/org/bukkit/event/entity/EntityPortalExitEvent.java:
```diff
@@ -3,25 +3,36 @@
  * <p>
  * This event allows you to modify the velocity of the entity after they
- * have successfully exeted the portal.
+ * have successfully exited the portal.
```
~~Sorry, but this belongs in the https://github.com/Bukkit/Bukkit-Javadoc/ repository.~~

turt2live replied:
This may be allowed to stay...

turt2live commented on src/main/java/org/bukkit/event/entity/EntityTeleportEvent.java:
```diff
@@ -22,6 +28,45 @@ public EntityTeleportEvent(Entity what, Location from, Location to) {
+    	return cause;
+    }
+    
+    public enum TeleportCause {
```
What's wrong with the existing cause?

Obliv replied:
Nothing. I'll remove this.

turt2live commented:
@Obliv This will need testing materials and the above issues addressed. 

As well, where is the implementation (CraftBukkit) side of this?

Obliv commented:
I will implement this as soon as possible, but that may be a while (probably the weekend). Feel free to close this request, if you think it's necessary.

Obliv committed:
Used existing teleport cause, removed javadocs

riking commented on src/main/java/org/bukkit/event/entity/EntityTeleportEvent.java:
```diff
@@ -14,7 +15,9 @@
 
+    @Deprecated
     public EntityTeleportEvent(Entity what, Location from, Location to) {
         super(what);
```
This constructor should call `this(what, from, to, TeleportCause.UNKNOWN)` I believe.

turt2live replied:
^

turt2live commented:
Currently this PR is missing the following:
- An implementation (I could not find any)
- Some sort of testing materials just to prove this API works
- The concerns outlined in the diff of this PR would need to be addressed (pretty minor)

If anyone can tackle these tasks, the chance of this PR getting pulled significantly increases.

Ribesg commented:
I think that this PR can be closed.

Wolvereness commented:
Abandoned.
